core/util/recover.go

```diff
@@ -0,0 +1,14 @@
+package util
+
+func WithRecover(fn func()) (success bool) {
+	defer func() {
+		if recover() != nil {
+			//recover panic from `send on closed channel`
```

Comment:
It's kind of hacky to use recover() for this, instead, we should organize code better to avoid sending on closed channel. The principle is never closed the channel from the reader end, or close only when all writers have finish their work, which can easily achived by using the sync.WaitGroup.

Reply:
removed